src/client/kv_client.rs

```diff
@@ -27,49 +31,94 @@ impl KvResponse {
     pub fn cas(&self) -> u64 {
         self.cas
     }
+
+    pub fn id(&self) -> String {
+        self.key.clone()
+    }
 }
 
-// Thinking here that some of this will need to go into arc mutexes at some point.
 pub struct KvClient {
     seeds: Vec<String>,
-    username: String,
-    password: String,
     manifest: Option<CollectionManifest>,
     endpoints: HashMap<String, KvEndpoint>,
-    config: Option<BucketConfig>,
+    config: BucketConfig,
     tls_config: ClusterTlsConfig,
     http_agent: HTTPHandler,
+    bucket: String,
 }
 
 impl KvClient {
-    pub fn new(
+    pub async fn connect(
         seeds: Vec<String>,
         username: String,
         password: String,
         tls_config: ClusterTlsConfig,
-    ) -> Self {
-        Self {
+        bucket: String,
+        deadline: Instant,
+        ctrl_c: Arc<AtomicBool>,
+    ) -> Result<Self, ClientError> {
+        let now = Instant::now();
+        if now >= deadline {
+            return Err(ClientError::Timeout { key: None });
+        }
+        let deadline_sleep = sleep(deadline.sub(now));
+        tokio::pin!(deadline_sleep);
+
+        let ctrl_c_fut = CtrlcFuture::new(ctrl_c.clone());
+        tokio::pin!(ctrl_c_fut);
+
+        let http_agent = HTTPHandler::new(username.clone(), password.clone(), tls_config.clone());
+        let config = KvClient::get_bucket_config(
+            seeds.clone(),
+            bucket.clone(),
+            tls_config.clone(),
+            &http_agent,
+            deadline,
+            ctrl_c.clone(),
+        )
+        .await?;
+
+        let mut endpoints = HashMap::new();
+        for addr in config.key_value_seeds(tls_config.enabled()) {
+            let connect = KvEndpoint::connect(
+                addr.0.clone(),
+                addr.1,
+                username.clone(),
+                password.clone(),
+                bucket.clone(),
+                tls_config.clone(),
+            );
+
+            let endpoint = select! {
+                res = connect => res,
+                () = &mut deadline_sleep => Err(ClientError::Timeout{key: None}),
+                () = &mut ctrl_c_fut => Err(ClientError::Cancelled{key: None}),
+            }?;
```

Comment:
since this will back out if one endpoint times out - what happens to the ones already connected? does the Drop handle the disconnect for them automatically?

Reply:
I've not actually looked into if this will terminate the connection but I would assume that the connection future will stop being polled and it'll drop the KvEndpoint triggering disconnect.